### PR TITLE
certgen: Add initial version

### PR DESCRIPTION
**Pull request by gandro**
This adds the initial version of the `cilium-certgen` tool. This initial
version is able to generate and store the Hubble CA, Hubble server, and
Hubble Relay client and server certificates. All options can either be
configured via command-line options or environment variables.

The actual deployment will happen via the Kubernetes manifests generated
by the Cilium Helm charts.

### CODE REVIEW

File: cmd/certgen.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+			log.Infof("%s %s", binaryName, version.Version)
+
+			if err := generateCertificates(); err != nil {
+				log.WithError(err).Fatal("failed to generate certificates")
+			}
+		},
+	}
+	rootCmd.SetVersionTemplate("{{with .Name}}{{printf \"%s \" .}}{{end}}{{printf \"v%s\" .Version}}\n")
+
+	flags := rootCmd.PersistentFlags()
```
Comment (Rolinh) [inline cmd/certgen.go:64]: nit: there are no sub-commands so these flags to need to be persistent.
```suggestion
	flags := rootCmd.Flags()
```

File: cmd/certgen.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+		log.Info("Loading Hubble CA from file")
+		err = hubbleCA.LoadFromFile(option.Config.HubbleCACertFile, option.Config.HubbleCAKeyFile)
+		if err != nil {
+			return fmt.Errorf("failed to load Hubble CA: %w", err)
+		}
+	}
+
+	var hubbleServerCert *generate.Cert
+	if option.Config.HubbleServerCertGenerate {
+		log.Info("Hubble server certificate generation enabled")
```
Comment (Rolinh) [inline cmd/certgen.go:152]: nit:
```suggestion
		log.Info("Generating certificates for Hubble server")
```

File: cmd/certgen.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		)
+		err := hubbleServerCert.Generate(hubbleCA.CACert, hubbleCA.CAKey)
+		if err != nil {
+			return fmt.Errorf("failed to generate Hubble server cert: %w", err)
+		}
+	}
+
+	var hubbleRelayClientCert *generate.Cert
+	if option.Config.HubbleRelayClientCertGenerate {
+		log.Info("Hubble Relay client certificate generation enabled")
```
Comment (Rolinh) [inline cmd/certgen.go:168]: nit: ditto

File: cmd/certgen.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+		)
+		err := hubbleRelayClientCert.Generate(hubbleCA.CACert, hubbleCA.CAKey)
+		if err != nil {
+			return fmt.Errorf("failed to generate Hubble Relay client cert: %w", err)
+		}
+	}
+
+	var hubbleRelayServerCert *generate.Cert
+	if option.Config.HubbleRelayServerCertGenerate {
+		log.Info("Hubble Relay server certificate generation enabled")
```
Comment (Rolinh) [inline cmd/certgen.go:184]: nit: ditto

File: internal/generate/generate.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+			Namespace: c.K8sSecretNamespace,
+		},
+		Data: map[string][]byte{
+			"tls.crt": c.CertBytes,
+			"tls.key": c.KeyBytes,
+		},
+		Type: v1.SecretTypeTLS,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.K8sRequestTimeout)
```
Comment (Rolinh) [inline internal/generate/generate.go:143]: nit: for future-proofing, we should probably change the signature of `StoreAsSecret` to take `ctx context.Context` as a parameter.

File: internal/generate/generate.go [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      c.K8sConfigMapName,
+			Namespace: c.K8sConfigMapNamespace,
+		},
+		BinaryData: map[string][]byte{
+			"ca.crt": c.CACertBytes,
+		},
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.K8sRequestTimeout)
```
Comment (Rolinh) [inline internal/generate/generate.go:258]: not: same comment as above: pass `ctx` as parameter to `StoreAsConfigMap`?

**Review by Rolinh — APPROVED** [review]

File: internal/generate/generate.go (hunk 5 above)
Comment (gandro) [reply internal/generate/generate.go:143]: Good idea. It also removes the dependency on the `option` package, which was unintended actually.

**Pull request merged**